aten/src/ATen/core/autocast/autocast_mode.h

```diff
@@ -0,0 +1,15 @@
+#pragma once
+
+#include <c10/macros/Macros.h>
+
+namespace at {
+namespace autocast {
+
+struct CAFFE2_API AutocastMode {
```

Comment:
Thinking out loud; it might be better to encourage people to only use the RAII guard version. That version would hopefully handle nested autocasting and exceptions correctly. (This is fine as an implementation detail, just try to avoid using it as the full interface)

Reply:
The Python exposure is purely via context manager, at least.

Reply:
A C++-side RAII guard doesn't really cover our use case, unfortunately, because the autocasting state needs to span multiple calls from the Python into the C++.